### PR TITLE
Add an 'Exception' module to take care of catching exceptions with a trace

### PR DESCRIPTION
**Pull request by mjambon**
In semgrep-core, we had a number of places where a stack trace was lost or collected at the wrong time. This is meant to enforce best practices.

See explanations in the new ml/mli files.

### Security

- [x] Change has no security implications (otherwise, ping the security team)

### CODE REVIEW

File: commons/Exception.mli [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+   1. Raise fresh exceptions normally using 'raise'.
+   2. Use 'Exception.reraise' to reraise exceptions.
+   3. Right after catching an exception with a try-with,
+      call 'Exception.catch'.
+   4. When defining a new exception, define an exception printer and
+      register it globally with 'Printexc.register_printer'.
+
+   * Do not pass exceptions around without an accompanying trace.
+   * Do not re-raise exceptions using 'raise'.
```
Comment (aryx) [inline commons/Exception.mli:24]: We might want to write a semgrep rule for that?

**Review by aryx — APPROVED** [review]
Impeccable.

File: commons/Exception.mli (hunk 1 above)
Comment (mjambon) [reply commons/Exception.mli:24]: I did that, found a lot more (correct) matches than I was expecting, and pushed the rule to the public repository of Semgrep rules for the public benefit: https://github.com/returntocorp/semgrep-rules/pull/2142
This Semgrep thing is actually useful.

Comment (aryx) [reply commons/Exception.mli:24]: haha

**Pull request merged**